### PR TITLE
drivers: modem_cellular: add Sequans GM02S modem

### PR DESCRIPTION
**Pull request by vThibo**
This pull requests adds the Sequans Monarch 2 GM02S modem family to the modem_cellular driver. Tested on the DPTechnics Walter board.

The GM02S modem returns `NO CARRIER` if it isn't registered yet before executing `ATD*99***1#`. Currently, a delay is added after `CFUN=1` to give the modem some time to register. If it didn't manage to register in time, the dial script fails.
Feedback is welcome if there's interest in changing the driver to add a new, or move the existing `AWAIT_REGISTRATION` state before executing the dial command.

### CODE REVIEW

**Review by bjarki-andreasen — APPROVED** [review]
Looks good :)

File: drivers/modem/modem_cellular.c [hunk 1]
```diff
@@ -2038,6 +2038,45 @@ MODEM_CHAT_SCRIPT_DEFINE(nordic_nrf91_slm_periodic_chat_script,
...
+MODEM_CHAT_SCRIPT_DEFINE(sqn_gm02s_init_chat_script, sqn_gm02s_init_chat_script_cmds,
+			 abort_matches, modem_cellular_chat_callback_handler, 10);
+
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(sqn_gm02s_dial_chat_script_cmds,
+			      MODEM_CHAT_SCRIPT_CMD_RESP_MULT("AT+CGACT=0,1", allow_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CGDCONT=1,\"IP\","
+							 "\""CONFIG_MODEM_CELLULAR_APN"\"",
+							 ok_match),
+				  MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT+CFUN=1", 10000),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("ATD*99***1#", connect_match));
```
Comment (bjarki-trackunit) [inline drivers/modem/modem_cellular.c:2067]: You can add an "empty" / "noop" script command here to validate the response to `AT+CFUN=1`, so it looks like
```
MODEM_CHAT_SCRIPT_CMD_RESP("AT+CFUN=1", ok_match),
MODEM_CHAT_SCRIPT_CMD_RESP_NONE("", 10000),
```
this way, the script fails if "OK" is not returned, while still waiting 10 seconds after :)

**Review by bjarki-andreasen — DISMISSED** [review]

**Review by rerickson1 — DISMISSED** [review]

File: drivers/modem/modem_cellular.c (hunk 1 above)
Comment (vThibo) [reply drivers/modem/modem_cellular.c:2067]: Thanks for the suggestion.

If I leave the command empty: `MODEM_CHAT_SCRIPT_CMD_RESP_NONE("", 10000),`, the driver does not run the dial script.

```
[00:00:03.360,000] <dbg> modem_cellular: modem_cellular_log_event: event dlci2 opened
[00:00:03.360,000] <dbg> modem_cellular: modem_cellular_log_state_changed: switch from open dlci2 to run dial script
[00:00:03.460,000] <dbg> modem_cellular: modem_cellular_log_event: event timeout
L4 was not connected in time
```

However adding a space: `MODEM_CHAT_SCRIPT_CMD_RESP_NONE(" ", 10000),` or a random AT command:`MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 10000),` does work normally.

```
[00:00:03.643,000] <dbg> modem_cellular: modem_cellular_log_event: event dlci2 opened
[00:00:03.643,000] <dbg> modem_cellular: modem_cellular_log_state_changed: switch from open dlci2 to run dial script
[00:00:03.743,000] <dbg> modem_cellular: modem_cellular_log_event: event timeout
[00:00:03.743,000] <dbg> modem_chat: modem_chat_script_start: running script: sqn_gm02s_dial_chat_script
[00:00:03.743,000] <dbg> modem_chat: modem_chat_script_next: sqn_gm02s_dial_chat_script: step: 0
[00:00:03.743,000] <dbg> modem_chat: modem_chat_script_next: sending: AT+CGACT=0,1
```

Comment (vThibo) [reply drivers/modem/modem_cellular.c:2067]: In modem_chat_run_script_async() of modem_chat.c, the script commands are validated:

```c
	/* Validate script commands */
	for (uint16_t i = 0; i < script->script_chats_size; i++) {
		if ((script->script_chats[i].request_size == 0) &&
		    (script->script_chats[i].response_matches_size == 0)) {
			return -EINVAL;
		}
	}
```

In the case of `MODEM_CHAT_SCRIPT_CMD_RESP_NONE("", 10000),`, both the request_size and matches_size is zero.

Comment (bjarki-trackunit) [reply drivers/modem/modem_cellular.c:2067]: Could you try updating the code to
```
	for (uint16_t i = 0; i < script->script_chats_size; i++) {
		if ((script->script_chats[i].request_size == 0) &&
		    (script->script_chats[i].response_matches_size == 0) &&
		    (script->script_chats[i].timeout == 0)) {
			return -EINVAL;
		}
	}
```

Comment (vThibo) [reply drivers/modem/modem_cellular.c:2067]: Works great! Do I add this as a new commit to the PR?

Comment (bjarki-trackunit) [reply drivers/modem/modem_cellular.c:2067]: Nah :) I will add the fix along with a test in the modem_chat test suite to make sure it stays fixed, then we can update the modem_cellular driver :)

Comment (bjarki-trackunit) [reply drivers/modem/modem_cellular.c]: PR for fixing `modem_chat` is here https://github.com/zephyrproject-rtos/zephyr/pull/72280

Also, you latest commit is using 
```
MODEM_CHAT_SCRIPT_CMD_RESP("AT+CFUN=1", ok_match),
MODEM_CHAT_SCRIPT_CMD_RESP_NONE("", 10000),
```
should be 
```
MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT+CFUN=1", 10000),
```
for now, or maybe I'm lost :)

**Review by bjarki-andreasen — APPROVED** [review]
Perfect :)

**Review by rerickson1 — APPROVED** [review]

**Pull request merged**